File: game/src/bullet.rs
//! Game project.
use fyrox::asset::Resource;
use fyrox::core::ComponentProvider;
use fyrox::resource::model::Model;
use fyrox::{
    core::{
        algebra::Vector3, pool::Handle, reflect::prelude::*, type_traits::prelude::*,
        visitor::prelude::*, TypeUuidProvider,
    },
    scene::node::Node,
    script::{ScriptContext, ScriptTrait},
};
use fyrox_lite_api::lite_ctx::{LiteContext, LiteScript};
use fyrox_lite_api::lite_math::{LiteQuaternion, LiteVector3};
use fyrox_lite_api::lite_node::LiteRoutingStrategy;
use fyrox_lite_api::lite_physics::{LitePhysics, LiteRayCastOptions};
use fyrox_lite_api::lite_prefab::LitePrefab;

#[derive(Visit, Reflect, Debug, Clone, TypeUuidProvider, ComponentProvider, Default)]
#[type_uuid(id = "12371d19-9f1a-4286-8486-add4ebaadaec")]
#[visit(optional)]
pub struct Bullet {
    pub params: BulletParams,
}

#[derive(Visit, Reflect, Debug, Clone, Default)]
pub struct BulletParams {
    pub velocity: Vector3<f32>,
    pub remaining_sec: f32,
    pub author_collider: Handle<Node>,
}

pub struct BulletSeed {
    pub prefab: Resource<Model>,
    pub origin: LiteVector3,
    pub direction: LiteVector3,
    pub initial_velocity: f32,
    pub author_collider: Handle<Node>,
    pub range: f32,
}

impl Bullet {
    pub fn spawn(seed: BulletSeed) {
        let orientation = LiteQuaternion::face_towards(seed.direction, LiteVector3::y_axis());
        let bullet = LitePrefab::new(seed.prefab).instantiate_at(seed.origin, orientation);
        bullet.with_script::<Bullet>(|it| {
            it.params = BulletParams {
                velocity: seed.direction.normalize().mul(seed.initial_velocity).into(),
                remaining_sec: seed.range / seed.initial_velocity,
                author_collider: seed.author_collider,
            };
        });
    }
}

#[derive(Debug)]
pub struct BulletHit {}

impl LiteScript for Bullet {
    fn on_update(&mut self, ctx: &mut LiteContext) {
        profiling::scope!("Bullet::on_update");
        self.params.remaining_sec -= ctx.dt;
        if self.params.remaining_sec <= 0.0 {
            ctx.node.destroy();
            return;
        }
        let new_pos = ctx.node.local_position().add(LiteVector3::from(self.params.velocity).mul(ctx.dt));

        let opts = LiteRayCastOptions {
            ray_origin: ctx.node.local_position(),
            ray_direction: LiteVector3(self.params.velocity).normalize(),
            max_len: self.params.velocity.magnitude() * ctx.dt,
            groups: Default::default(),
            sort_results: true,
        };
        let mut results = vec![];
        {
            profiling::scope!("cast_ray");
            LitePhysics::cast_ray(opts, &mut results);
        }

        for hit in results {
            if hit.collider == self.params.author_collider.into() {
                continue;
            }
            hit.collider
                .send_hierarchical(LiteRoutingStrategy::Up, BulletHit {});
            ctx.node.destroy();
            return;
        }

        ctx.node.set_local_position(new_pos);
    }
}

impl ScriptTrait for Bullet {
    fn on_update(&mut self, ctx: &mut ScriptContext) {
        self.redispatch_update(ctx);
    }
}

File: scripts/Bullet.lua
---@uuid 12371d19-9f1a-4286-8486-add4ebaadaec
---@class Bullet : Script
---@field velocity Vector3
---@field remaining_sec number
---@field author_collider Node
Bullet = script_class()

Bullet.HitMessage = "BulletHit"


---@class BulletSeed
---@field prefab Prefab
---@field origin Vector3
---@field direction Vector3
---@field initial_velocity number
---@field author_collider Node
---@field range number
local BulletSeed = {}

---@param scene Scene
---@param seed BulletSeed
function Bullet:spawn(seed)
    local orientation = Quaternion:face_towards(seed.direction, Vector3.Y);
    local bullet = seed.prefab:instantiate_at(seed.origin, orientation);
    local script = bullet:add_script("Bullet")
    script.velocity = seed.direction:normalize():mul(seed.initial_velocity)
    script.remaining_sec = seed.range / seed.initial_velocity
    script.author_collider = seed.author_collider
end

---@param ctx ScriptContext
function Bullet:on_update(dt)
    self.remaining_sec = self.remaining_sec - dt
    if self.remaining_sec <= 0.0 then
        ctx.handle:remove_node()
        return;
    end
    local new_pos = self.node:local_position():add(self.velocity:mul(dt))

    ---@type RayCastOptions
    local opts = {
        ray_origin= self.node:local_position(),
        ray_direction= self.velocity:normalize(),
        max_len=self.velocity:magnitude() * dt,
        sort_results= true
    }
    local results = {}
    Physics:cast_ray(opts, results)

    for i, hit in ipairs(results) do
        if hit.collider ~= self.author_collider then
            hit.collider:send_hierarchical(RoutingStrategy.Up, Bullet.HitMessage)
			self.node:destroy()
            return;
        end
    end

    self.node:set_local_position(new_pos);
end